File: rust-lib-async/src/conn.rs
//! Per-connection handling: accepting, connecting (with reconnection), the
//! HELLO handshake, and the reader/writer task pair that pumps one TCP
//! connection. All data flows to and from the [`crate::broker`] over its
//! command channel.

use std::time::Duration;

use bytes::Bytes;
use tokio::io::{AsyncRead, AsyncWrite};
use tokio::net::{TcpListener, TcpStream, ToSocketAddrs};
use tokio::sync::{mpsc, oneshot};
use tokio::time::timeout;
use tracing::{debug, warn};

use crate::broker::Command;
use crate::protocol::{self, Envelope, Identity};
use crate::Shutdown;

const HEARTBEAT_INTERVAL: Duration = Duration::from_secs(5);
const HEARTBEAT_TIMEOUT: Duration = Duration::from_secs(15);
const CONNECT_TIMEOUT: Duration = Duration::from_secs(1);

type ReconnectDelay = std::sync::Arc<dyn Fn() -> Duration + Send + Sync>;

/// How an accepted TCP stream becomes the byte stream we speak the protocol
/// over: directly (`Plain`) or after a TLS server handshake (`Tls`). This is the
/// one and only place the bind side knows about TLS; the rest of the connection
/// machinery is generic over the resulting stream.
#[derive(Clone)]
pub(crate) enum Acceptor {
    Plain,
    #[cfg(feature = "tls")]
    Tls(tokio_rustls::TlsAcceptor),
}

/// The connect-side counterpart of [`Acceptor`]: dial plainly, or perform a TLS
/// client handshake against the carried server name.
#[derive(Clone)]
pub(crate) enum Connector {
    Plain,
    #[cfg(feature = "tls")]
    Tls(
        tokio_rustls::TlsConnector,
        tokio_rustls::rustls::pki_types::ServerName<'static>,
    ),
}

/// Accept connections until shutdown, spawning a handler per peer.
pub(crate) async fn accept_loop(
    listener: TcpListener,
    acceptor: Acceptor,
    cmd_tx: mpsc::UnboundedSender<Command>,
    identity: Identity,
    mut shutdown: Shutdown,
) {
    loop {
        tokio::select! {
            _ = shutdown.recv() => break,
            res = listener.accept() => match res {
                Ok((stream, _addr)) => {
                    let _ = stream.set_nodelay(true);
                    let acceptor = acceptor.clone();
                    let cmd_tx = cmd_tx.clone();
                    let sd = shutdown.clone();
                    tokio::spawn(async move {
                        let res = match acceptor {
                            Acceptor::Plain => handle_connection(stream, cmd_tx, identity, sd).await,
                            #[cfg(feature = "tls")]
                            Acceptor::Tls(a) => match a.accept(stream).await {
                                Ok(tls) => handle_connection(tls, cmd_tx, identity, sd).await,
                                Err(e) => { debug!("TLS server handshake failed: {e}"); Ok(()) }
                            },
                        };
                        if let Err(e) = res {
                            debug!("inbound connection ended: {e}");
                        }
                    });
                }
                Err(e) => warn!("accept error: {e}"),
            },
        }
    }
}

/// Connect to `addr` and keep reconnecting (after `delay()`) for the life of
/// the socket.
pub(crate) async fn connect_loop<A>(
    addr: A,
    connector: Connector,
    cmd_tx: mpsc::UnboundedSender<Command>,
    identity: Identity,
    delay: ReconnectDelay,
    mut shutdown: Shutdown,
) where
    A: ToSocketAddrs + Clone,
{
    while !shutdown.is_shutdown() {
        let stream = tokio::select! {
            _ = shutdown.recv() => break,
            res = timeout(CONNECT_TIMEOUT, TcpStream::connect(addr.clone())) => match res {
                Ok(Ok(s)) => Some(s),
                Ok(Err(e)) => { debug!("connect failed: {e}"); None }
                Err(_) => { debug!("connect timed out"); None }
            },
        };

        if let Some(stream) = stream {
            let _ = stream.set_nodelay(true);
            let res = match &connector {
                Connector::Plain => {
                    handle_connection(stream, cmd_tx.clone(), identity, shutdown.clone()).await
                }
                #[cfg(feature = "tls")]
                Connector::Tls(c, name) => match c.connect(name.clone(), stream).await {
                    Ok(tls) => {
                        handle_connection(tls, cmd_tx.clone(), identity, shutdown.clone()).await
                    }
                    Err(e) => {
                        debug!("TLS client handshake failed: {e}");
                        Ok(())
                    }
                },
            };
            if let Err(e) = res {
                debug!("outbound connection ended: {e}");
            }
        }

        if shutdown.is_shutdown() {
            break;
        }
        let wait = (delay.as_ref())();
        tokio::select! {
            _ = shutdown.recv() => break,
            _ = tokio::time::sleep(wait) => {}
        }
    }
}

/// Drive one established connection: handshake, register with the broker, then
/// run its reader and writer until either side ends. Generic over the byte
/// stream so plain TCP and TLS-wrapped connections share one implementation.
async fn handle_connection<S>(
    stream: S,
    cmd_tx: mpsc::UnboundedSender<Command>,
    my_identity: Identity,
    mut shutdown: Shutdown,
) -> std::io::Result<()>
where
    S: AsyncRead + AsyncWrite + Unpin + Send + 'static,
{
    let (mut read_half, mut write_half) = tokio::io::split(stream);

    // --- Handshake: send our HELLO, read and validate the peer's. ----------
    protocol::write_frame(&mut write_half, &protocol::hello(&my_identity)).await?;
    let peer_identity = match protocol::read_frame(&mut read_half).await? {
        None => return Ok(()), // closed during handshake
        Some(frame) => match protocol::decode(&frame) {
            Some(Envelope::Hello { version, identity }) => {
                if version != protocol::PROTOCOL_VERSION {
                    warn!("unsupported protocol version {version}; closing");
                    return Ok(());
                }
                identity
            }
            _ => {
                warn!("expected HELLO handshake; closing");
                return Ok(());
            }
        },
    };

    // --- Register with the broker. ------------------------------------------
    let (writer_tx, mut writer_rx) = mpsc::unbounded_channel::<Bytes>();
    let (accept_tx, accept_rx) = oneshot::channel();
    if cmd_tx
        .send(Command::ConnectionUp {
            identity: peer_identity,
            writer: writer_tx,
            accept: accept_tx,
        })
        .is_err()
    {
        return Ok(()); // broker is gone
    }
    if !matches!(accept_rx.await, Ok(true)) {
        debug!("connection rejected (duplicate identity); closing");
        return Ok(());
    }

    // --- Writer task: drains the broker's queue, beating on idle. -----------
    let mut writer_shutdown = shutdown.clone();
    let writer = tokio::spawn(async move {
        loop {
            tokio::select! {
                _ = writer_shutdown.recv() => break,
                maybe = writer_rx.recv() => match maybe {
                    Some(envelope) => {
                        if protocol::write_frame(&mut write_half, &envelope).await.is_err() {
                            break;
                        }
                    }
                    None => break, // broker dropped this connection
                },
                _ = tokio::time::sleep(HEARTBEAT_INTERVAL) => {
                    if protocol::write_frame(&mut write_half, &protocol::heartbeat()).await.is_err() {
                        break;
                    }
                }
            }
        }
    });

    // --- Reader loop: forwards data envelopes; dies on heartbeat timeout. ---
    loop {
        tokio::select! {
            _ = shutdown.recv() => break,
            res = timeout(HEARTBEAT_TIMEOUT, protocol::read_frame(&mut read_half)) => match res {
                Err(_) => { debug!("heartbeat timeout; closing connection"); break; }
                Ok(Ok(None)) => break,   // clean close
                Ok(Err(_)) => break,     // io error
                Ok(Ok(Some(frame))) => match protocol::decode(&frame) {
                    // Heartbeats only reset the timeout; HELLO is unexpected
                    // here; unknown types are ignored.
                    Some(Envelope::Heartbeat) | Some(Envelope::Hello { .. }) | None => continue,
                    Some(envelope) => {
                        if cmd_tx
                            .send(Command::Received { identity: peer_identity, envelope })
                            .is_err()
                        {
                            break;
                        }
                    }
                },
            },
        }
    }

    let _ = cmd_tx.send(Command::ConnectionDown {
        identity: peer_identity,
    });
    writer.abort();
    let _ = writer.await;
    Ok(())
}

File: golang-lib/conn.go
package aiomsg

// Per-connection handling: accepting, connecting (with reconnection), the HELLO
// handshake, and the reader/writer goroutine pair that pumps one TCP
// connection. All data flows to and from the broker over its command channel.

import (
	"context"
	"crypto/tls"
	"net"
	"sync"
	"time"
)

const (
	heartbeatInterval = 5 * time.Second
	heartbeatTimeout  = 15 * time.Second
	connectTimeout    = 1 * time.Second
)

func (s *Socket) acceptLoop(ln net.Listener) {
	defer s.wg.Done()
	// Close the listener on shutdown so Accept returns.
	go func() {
		<-s.ctx.Done()
		_ = ln.Close()
	}()
	for {
		conn, err := ln.Accept()
		if err != nil {
			return // listener closed (shutdown) or fatal accept error
		}
		s.wg.Add(1)
		go func(c net.Conn) {
			defer s.wg.Done()
			// Bind-side accept: sniff raw-vs-WebSocket (PROTOCOL.md §10) before
			// running the ordinary connection handler. The connect end never
			// receives WebSocket, so connectLoop calls handleConnection directly.
			wrapped, err := sniffAndWrap(c)
			if err != nil {
				_ = c.Close()
				return
			}
			s.handleConnection(wrapped)
		}(conn)
	}
}

func (s *Socket) connectLoop(addr string, tlsConfig *tls.Config) {
	defer s.wg.Done()
	for {
		select {
		case <-s.ctx.Done():
			return
		default:
		}

		d := net.Dialer{Timeout: connectTimeout}
		var conn net.Conn
		var err error
		if tlsConfig != nil {
			// tls.Dialer completes the TLS handshake as part of dialing.
			conn, err = (&tls.Dialer{NetDialer: &d, Config: tlsConfig}).DialContext(s.ctx, "tcp", addr)
		} else {
			conn, err = d.DialContext(s.ctx, "tcp", addr)
		}
		if err == nil {
			s.handleConnection(conn)
		}

		select {
		case <-s.ctx.Done():
			return
		case <-time.After(s.reconnectDelay()):
		}
	}
}

// handleConnection drives one established connection: handshake, register with
// the broker, then run reader and writer until either side ends.
func (s *Socket) handleConnection(conn net.Conn) {
	defer conn.Close()
	setNoDelay(conn)

	// Handshake: send our HELLO, read and validate the peer's.
	if err := writeFrame(conn, encodeHello(s.identity)); err != nil {
		return
	}
	_ = conn.SetReadDeadline(time.Now().Add(heartbeatTimeout))
	frame, err := readFrame(conn)
	if err != nil {
		return
	}
	env, ok := decode(frame)
	if !ok || env.Type != typeHello || env.Version != protocolVersion {
		return
	}
	peerID := env.Identity

	// Register with the broker.
	writerCh := make(chan []byte, writerBuffer)
	accept := make(chan bool, 1)
	select {
	case s.broker.commands <- command{kind: cmdConnUp, identity: peerID, writer: writerCh, accept: accept}:
	case <-s.ctx.Done():
		return
	}
	var accepted bool
	select {
	case accepted = <-accept:
	case <-s.ctx.Done():
		return
	}
	if !accepted {
		return // duplicate identity
	}

	connCtx, connCancel := context.WithCancel(s.ctx)
	defer connCancel()
	// Closing the conn when this connection's context is cancelled unblocks the
	// reader promptly on shutdown.
	go func() {
		<-connCtx.Done()
		_ = conn.Close()
	}()

	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		defer wg.Done()
		runWriter(connCtx, conn, writerCh)
	}()

	// Reader runs inline until the connection ends.
	runReader(connCtx, conn, peerID, s.broker.commands)

	// Tell the broker the connection is gone (it closes writerCh), stop writer.
	select {
	case s.broker.commands <- command{kind: cmdConnDown, identity: peerID}:
	case <-s.ctx.Done():
	}
	connCancel()
	wg.Wait()
}

func runReader(ctx context.Context, conn net.Conn, peerID Identity, commands chan command) {
	for {
		_ = conn.SetReadDeadline(time.Now().Add(heartbeatTimeout))
		frame, err := readFrame(conn)
		if err != nil {
			return // heartbeat timeout or connection closed
		}
		env, ok := decode(frame)
		if !ok {
			continue // unknown envelope; ignore
		}
		switch env.Type {
		case typeHeartbeat, typeHello:
			continue // heartbeats only reset the deadline; HELLO unexpected here
		default:
			select {
			case commands <- command{kind: cmdReceived, identity: peerID, env: env}:
			case <-ctx.Done():
				return
			}
		}
	}
}

func runWriter(ctx context.Context, conn net.Conn, writerCh chan []byte) {
	timer := time.NewTimer(heartbeatInterval)
	defer timer.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case env, ok := <-writerCh:
			if !ok {
				return // broker dropped this connection
			}
			if err := writeFrame(conn, env); err != nil {
				return
			}
			resetTimer(timer, heartbeatInterval)
		case <-timer.C:
			if err := writeFrame(conn, encodeHeartbeat()); err != nil {
				return
			}
			timer.Reset(heartbeatInterval)
		}
	}
}

// setNoDelay disables Nagle's algorithm on the connection's underlying TCP
// socket, reaching through a *tls.Conn if necessary.
func setNoDelay(conn net.Conn) {
	if tc, ok := conn.(*tls.Conn); ok {
		conn = tc.NetConn()
	}
	if tcp, ok := conn.(*net.TCPConn); ok {
		_ = tcp.SetNoDelay(true)
	}
}

func resetTimer(t *time.Timer, d time.Duration) {
	if !t.Stop() {
		select {
		case <-t.C:
		default:
		}
	}
	t.Reset(d)
}
